packages/core/contracts/bot-helpers/atomic-liquidator/ReserveCurrencyLiquidator.sol

```diff
@@ -0,0 +1,136 @@
+
+    /**
+     * @notice Swaps required amount of reserve currency to collateral currency which is then used to mint tokens to
+     * liquidate a position within one transaction.
+     * @dev After the liquidation is done the DSProxy that called this method will have an open positon AND pending
+     * liquidation within the financial contract. The bot using the DSProxy should withdraw the liquidation once it has
+     * passed liveness. At this point the position can be manually unwound.
+     * @param uniswapRouter address of the uniswap router used to facilate trades.
+     * @param financialContract address of the financial contract on which the liquidation is occuring.
+     * @param reserveCurrency address of the token to swap for collateral. THis is the common currency held by the DSProxy.
+     * @param liquidatedSponsor address of the sponsor to be liquidated.
+     * @param maxReserverTokenSpent maximum number of reserve tokens to spend in the trade. Bounds slippage.
+     * @param minCollateralPerTokenLiquidated abort the liquidation if the position's collateral per token is below this value.
+     * @param maxCollateralPerTokenLiquidated abort the liquidation if the position's collateral per token exceeds this value.
+     * @param maxTokensToLiquidate max number of tokens to liquidate. For a full liquidation this is the full position debt.
+     * @param deadline abort the trade and liquidation if the transaction is mined after this timestamp.
+     **/
+    function swapMintLiquidate(
+        address uniswapRouter,
+        address financialContract,
+        address reserveCurrency,
+        address liquidatedSponsor,
+        FixedPoint.Unsigned calldata maxReserverTokenSpent,
+        FixedPoint.Unsigned calldata minCollateralPerTokenLiquidated,
+        FixedPoint.Unsigned calldata maxCollateralPerTokenLiquidated,
+        FixedPoint.Unsigned calldata maxTokensToLiquidate,
+        uint256 deadline
+    ) public {
+        IUniswapV2Router01 router = IUniswapV2Router01(uniswapRouter);
+        FinancialContractInterface fc = FinancialContractInterface(financialContract);
+
+        // 1. Calculate how much collateral needed to mint maxTokensToLiquidate.
+        FixedPoint.Unsigned memory gcr = fc.pfc().div(fc.totalTokensOutstanding());
+        FixedPoint.Unsigned memory collateralToMintAtGcr = maxTokensToLiquidate.mul(gcr);
+
+        // 2. Calculate how much collateral is needed for the final fee.
+        FinderInterface finder = FinderInterface(fc.finder());
+        StoreInterface store = StoreInterface(finder.getImplementationAddress("Store"));
+        FixedPoint.Unsigned memory totalCollateralNeeded =
+            collateralToMintAtGcr.add(store.computeFinalFee(fc.collateralCurrency()));
+
+        // 3. Swap reserve currency to get required collateral for the mint + final fee.
+        address[] memory path = new address[](2);
+        path[0] = reserveCurrency;
+        path[1] = fc.collateralCurrency();
```

Comment:
Is the path valid if path[0] == path[1] (i.e. the same currency?)

Reply:
This is actually a great point. If the reserve currency is the collateral currency then we should not do any trade. It is posible that the uniswap router will gracefully deal with this but it might be better to have a bit of branching logic within the contract. I will add some unit tests to validate this and update the contracts accordingly if this needs to be changed.